### PR TITLE
ShaderMaterial: Improved serialization/deserialization

### LINKED ISSUE #10288: Add the possibility of loading textures with ShaderMaterials
##### Description of the problem 
I have a custom ShaderMaterial that requires a texture map and currently this doesn't seem to be supported. I would need the following code added to the MaterialLoader.js inside the parse function:
```js              
// for ShaderMaterials
if(material.uniforms !== undefined){ 

	for ( var u in material.uniforms ){

		if ( u === "textureMap" && material.uniforms[u].value ){

			material.uniforms[u].value = getTexture( material.uniforms[u].value );

		}

	}
}
```
I this this could be very useful and offer a lot of possibilities with custom shaders

##### Three.js version

- [ ] Dev
- [x] r82
- [ ] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] Linux
- [ ] Android
- [ ] IOS

### PR DESCRIPTION
**Pull request by Mugen87**
Fixed #10288

### CODE REVIEW

**Comment by Mugen87** [comment]
BTW: It's not necessary that the `type` property of uniforms is used. Texture are detected via `Texture.isTexture` by `ShaderMaterial.toJSON()`.

**Comment by Mugen87** [comment]
This PR actually fixes #10040

**Comment by SimonDanisch** [comment]
Any chance to merge this?! 
@rdeits can we update the threejs version of MeshCat if this happens?

File: src/materials/ShaderMaterial.js [hunk 1]
```diff
@@ -116,7 +116,31 @@ ShaderMaterial.prototype.toJSON = function ( meta ) {
...
+
+			data.uniforms[ name ] = {
+				type: 't',
+				value: texture.toJSON( meta ).uuid
+			};
+
+		} else {
+
+			data.uniforms[ name ] = {
+				value: uniform.value
```
Comment (takahirox) [inline src/materials/ShaderMaterial.js]: Don't we need `value` serialization here? I think `value` can be Three.js object, like `THREE.Color`, `THREE.Vector3`, `THREE.Matrix`, and so on.

File: src/loaders/MaterialLoader.js [hunk 2]
```diff
@@ -100,6 +97,24 @@ Object.assign( MaterialLoader.prototype, {
...
+		// Shader Material
+
+		if ( json.uniforms !== undefined ) {
+
+			for ( var name in json.uniforms ) {
+
+				var uniform = json.uniforms[ name ];
+
+				material.uniforms[ name ] = {};
+				material.uniforms[ name ].value = ( uniform.type === 't' ) ? getTexture( uniform.value ) : uniform.value;
```
Comment (takahirox) [inline src/loaders/MaterialLoader.js]: I think we need Three.js object deserialization here. `value` can be `THREE.Color`, `THREE.Vector3`, `THREE.Matrix`, or so on.

File: src/materials/ShaderMaterial.js (hunk 1 above)
Comment (Mugen87) [reply src/materials/ShaderMaterial.js]: Okay, let's expand the PR a bit. I will add support for common uniform types.

**Comment by Mugen87** [comment]
There is now support for `Texture`, `Color`, `Vector2`, `Vector3`, `Vector4` and `Matrix4`.

**Comment by takahirox** [comment]
What types this PR doesn't support yet? `CubeTexture`?

**Comment by Mugen87** [comment]
AFAIK, `CubeTexture` is not a valid uniform value (see https://github.com/mrdoob/three.js/wiki/Uniforms-types). What's still missing are the array variants  `v2v`, `v3v`, `v4v`, `m4v` and `tv`. These types are not frequently used so we can still add them if there is a concrete feature request.

**Comment by takahirox** [comment]
> What's still missing are the array variants v2v, v3v, v4v, m4v and tv. These types are not frequently used so we can still add them if there is a concrete feature request.

Can we add a note about it in the code?

**Comment by Mugen87** [comment]
Done! 👍

**Comment by takahirox** [comment]
Thanks. BTW, do we also [de]serialize `.defines`? I think another PR is ok.

**Review by takahirox — APPROVED** [review]

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**